File: 백준/Silver/11723. 집합/집합.cc
#include<iostream>
#include<algorithm>
using namespace std;

int state;

int main(void) {
	cin.tie(NULL)->sync_with_stdio(false);

	int M;
	cin >> M;

	while (M--) {
		string command;
		int x;

		cin >> command;

		if (command == "add") {
			cin >> x;
			state |= (1 << (x - 1));
		}
		else if (command == "remove") {
			cin >> x;
			state &= (~(1 << (x - 1)));
		}
		else if (command == "check") {
			cin >> x;
			cout << ((state >> (x - 1)) & 1) << '\n';
		}
		else if (command == "toggle") {
			cin >> x;
			state ^= (1 << (x - 1));
		}
		else if (command == "all") {
			state = 0xfffff;
		}
		else {
			state = 0;
		}
	}

	return 0;
}

File: 백준/Silver/11723. 집합/집합.java
//이걸 비트마스킹으로 어떻게 풀지

import java.io.*;
import java.util.*;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        StringBuilder sb = new StringBuilder();

        int N = Integer.parseInt(st.nextToken());
        int S = 0;

        for (int i = 0; i < N; i++) {
            st = new StringTokenizer(br.readLine());

            String command = st.nextToken();

            if (command.equals("all")) {
                S = (1 << 21) - 1;
            } else if (command.equals("empty")) {
                S = 0;
            } else {
                int M = Integer.parseInt(st.nextToken());
                if (command.equals("add")) {
                    S |= 1 << M;
                } else if (command.equals("remove")) {
                    S &= ~(1 << M);
                } else if (command.equals("check")) {
                    sb.append((S & (1 << M)) != 0 ? 1 : 0).append("\n");
                } else if (command.equals("toggle")) {
                    S ^= 1 << M;
                }
            }
        }

        System.out.println(sb.toString());
    }
}
